File: 3517-shortest-distance-after-road-addition-queries-i/shortest-distance-after-road-addition-queries-i.py
class Solution:
    def shortestDistanceAfterQueries(self, n: int, queries: List[List[int]]) -> List[int]:
        adj = [[i + 1] for i in range(n - 1)]

        def shortestPath(query):
            adj[query[0]].append(query[1])
            #bfs
            queue = deque([(0, 0)])
            visited = set([0])
            while queue:
                pos, l = queue.popleft()
                if pos == n-1:
                    return l
                for i in adj[pos]:
                    if i not in visited:
                        visited.add(i)
                        queue.append((i, l+1))
            
        result = []
        for query in queries:
            result.append(shortestPath(query))
        return result

            

File: 3517-shortest-distance-after-road-addition-queries-i/shortest-distance-after-road-addition-queries-i.cpp
class Solution {
public:
    vector<int> shortestDistanceAfterQueries(int n, vector<vector<int>>& queries) {
        vector<vector<int>> adj(n - 1);
        for (int i = 0; i < n - 1; i++)
            adj[i].push_back(i+1);
        
        vector<int> result;
        for (vector<int> query: queries){
            adj[query[0]].push_back(query[1]);
            unordered_set<int> visited;
            queue<pair<int,int>> q;
            q.push(pair<int,int>(0, 0));
            visited.insert(0);
            while (q.size() > 0){
                pair<int,int> temp = q.front();
                q.pop();
                if (temp.first == n - 1)
                {
                    result.push_back(temp.second);
                    break;
                }
                for (int i: adj[temp.first]){
                    if (!visited.contains(i)){
                        q.push(pair<int,int>(i, temp.second + 1));
                        visited.insert(i);
                    }
                }
            }
        }
        return result;
    }
};